### PR TITLE
Adding DecodedIntensityTable class

### PR DESCRIPTION
**Pull request by shanaxel42**
Introducing DecodedIntensityTable which is just a subclass of IntensityTable. I moved anything related to decoded spots into the DecodedIntensityTable class. This is mostly to reduce confusion in the codebase between the outputs of various spot finding and decoding methods. In most cases the very end of a pipeline will now be a DecodedIntensityTable.

### CODE REVIEW

File: starfish/core/codebook/codebook.py [hunk 1]
```diff
@@ -552,10 +553,11 @@ def decode_metric(
...
         # add empty metadata fields and return
         if intensities.sizes[Features.AXIS] == 0:
-            intensities[Features.TARGET] = (Features.AXIS, np.empty(0, dtype='U'))
-            intensities[Features.DISTANCE] = (Features.AXIS, np.empty(0, dtype=float))
-            intensities[Features.PASSES_THRESHOLDS] = (Features.AXIS, np.empty(0, dtype=bool))
-            return intensities
+            return DecodedIntensityTable.from_intensity_table(
+                intensities,
+                targets=(Features.AXIS, np.empty(0, dtype='U')),
+                distances=(Features.AXIS, np.empty(0, dtype=float)),
```
[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply starfish/core/codebook/codebook.py]: Fine with this suggestion but suggest not dropping precision: suggest using `np.float64` if changing.

File: starfish/core/intensity_table/decoded_intensity_table.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+            targets=(Features.AXIS, np.random.choice(list('ABCD'), size=20)),
+            distances=(Features.AXIS, np.random.rand(20)))
+
+    def assign_cell_ids(self, cell_ids: Tuple[str, np.ndarray]):
+        """
+        Assign the given cell_ids to the DecodedIntensityTable
+        """
+        self[Features.CELL_ID] = cell_ids
+
+    def to_decoded_spots(self) -> DecodedSpots:
```
[the remark this answers is not in the document]
Comment (shanaxel42) [reply starfish/core/intensity_table/decoded_intensity_table.py]: xarray already has a `to_dataframe()` method so we can't override that, how about `to_decoded_dataframe`?

File: starfish/core/intensity_table/decoded_intensity_table.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        """
+        if Features.TARGET not in self.coords.keys():
+            raise RuntimeError(
+                "Intensities must be decoded before a DecodedSpots table can be produced.")
+        df = self.to_features_dataframe()
+        pixel_coordinates = pd.Index([Axes.X, Axes.Y, Axes.ZPLANE])
+        df = df.drop(pixel_coordinates.intersection(df.columns), axis=1).drop(Features.AXIS, axis=1)
+        return DecodedSpots(df)
+
+    def to_mermaid(self, filename: str) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (shanaxel42) [reply starfish/core/intensity_table/decoded_intensity_table.py]: This is out of scope of this PR. Made #1559

File: starfish/core/intensity_table/decoded_intensity_table.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        -------
+        DecodedIntensityTable
+        """
+        intensities = DecodedIntensityTable(intensities)
+        return cls.from_intensity_table(
+            intensities,
+            targets=(Features.AXIS, np.random.choice(list('ABCD'), size=20)),
+            distances=(Features.AXIS, np.random.rand(20)))
+
+    def assign_cell_ids(self, cell_ids: Tuple[str, np.ndarray]):
```
[the remark this answers is not in the document]
Comment (shanaxel42) [reply starfish/core/intensity_table/decoded_intensity_table.py]: I'll remove this method and just assign targets directly in tests

File: starfish/core/codebook/codebook.py [hunk 5]
```diff
@@ -641,10 +642,11 @@ def _view_row_as_element(array: np.ndarray) -> np.ndarray:
...
```
Comment (ttung) [inline starfish/core/codebook/codebook.py]: np.float64 like above.

File: starfish/core/intensity_table/decoded_intensity_table.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    Axes,
+    Coordinates,
+    DecodedSpots,
+    Features,
+)
+
+
+class DecodedIntensityTable(IntensityTable):
+    """
+    DecodedIntensityTable is a container for spot or pixel features extracted from
```
Comment (ttung) [inline starfish/core/intensity_table/decoded_intensity_table.py]: unwrap as per 100-col convention, pls.

File: starfish/core/intensity_table/decoded_intensity_table.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        if passes_threshold:
+            intensities[Features.PASSES_THRESHOLDS] = passes_threshold
+        return intensities
+
+    def to_decoded_dataframe(self) -> DecodedSpots:
+        """
+        Generates a dataframe containing decoded spot information. Guaranteed to contain physical
+        spot coordinates (z, y, x) and gene target. Does not contain pixel coordinates.
+        """
+        if Features.TARGET not in self.coords.keys():
```
Comment (ttung) [inline starfish/core/intensity_table/decoded_intensity_table.py]: This is not necessary any more.

File: starfish/core/intensity_table/decoded_intensity_table.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            Name for compressed-gzipped MERMAID data file. Should end in '.csv.gz'.
+
+        Notes
+        ------
+        See also https://github.com/JEFworks/MERmaid
+
+        """
+
+        # verify the IntensityTable has been decoded
+        if Features.TARGET not in self.coords.keys():
```
Comment (ttung) [inline starfish/core/intensity_table/decoded_intensity_table.py]: This is also not necessary any more.

File: starfish/core/intensity_table/test/test_intensity_table_coords.py [hunk 9]
```diff
@@ -7,6 +7,7 @@
 
 from starfish.core.codebook.test.factories import codebook_array_factory
 from starfish.core.imagestack.test.factories import imagestack_with_coords_factory
+from starfish.core.intensity_table.test import factories
```
Comment (ttung) [inline starfish/core/intensity_table/test/test_intensity_table_coords.py]: This should be a relative import, i.e., `from .factories import synthetic_decoded_intenisty_table`

**Comment by ttung** [comment]
Oh, can you verify the docs generate correctly for this?  I think we need to link it in from the `.rst` files.

**Pull request merged**